odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,663 @@
+
+        See Also
+        --------
+        read_data
+        """
+        # Read all fields except data. We use an OrderedDict such that
+        # the order is the same as in `header_fields`. This makes it simple
+        # to write later on in the correct order, based only on `header`.
+        header = OrderedDict()
+        for field in self.header_fields:
+            # Get all the values from the dictionary
+            name = field['name']
+            if name == 'data':
+                continue
+            entry = {'description': field.get('description', '')}
+            offset_bytes = int(field['offset'])
+            entry['offset'] = offset_bytes
+            size_bytes = int(field['size'])
+            dtype = np.dtype(field['dtype'])
+            shape = field.get('dshape', -1)  # no-op by default
+
+            if size_bytes is None:
+                # Default if 'size' is omitted
+                num_elems = 1
+            else:
+                num_elems = size_bytes / dtype.itemsize
+
+            if size_bytes % dtype.itemsize:
+                raise RuntimeError(
+                    "field '{}': `size` {} and `dtype.itemsize` {} "
+                    " result in non-integer number of elements"
+                    "".format(name, size_bytes, dtype.itemsize))
+
+            # Create format string for struct module to unpack the binary
+            # data
+            if np.issubdtype(dtype, np.dtype('S')):
+                # Have conversion only for 'S1', so we need to translate
+                fmt = (str(int(num_elems) * dtype.itemsize) + 's')
+            else:
+                # Format character can be obtained as dtype.char
+                fmt = str(int(num_elems)) + dtype.char
+
+            if struct.calcsize(fmt) != size_bytes:
+                raise RuntimeError(
+                    "field '{}': format '{}' results in {} bytes, but "
+                    "`size` is {}"
+                    "".format(name, fmt, struct.calcsize(fmt), size_bytes))
+
+            self.file.seek(offset_bytes)
+            packed_value = self.file.read(size_bytes)
+
+            if np.issubdtype(dtype, np.dtype('S')):
+                # Bytestring type, decode instead of unpacking. Replace
+                # \x00 characters with whitespace so the final length is
+                # correct
+                packed_value = packed_value.replace(b'\x00', b' ')
+                value = np.fromiter(packed_value.decode().ljust(size_bytes),
+                                    dtype=dtype)
+                entry['value'] = value.reshape(shape)
+            else:
+                value = np.array(struct.unpack_from(fmt, packed_value),
+                                 dtype=dtype)
+                entry['value'] = value.reshape(shape)
+
+            header[name] = entry
+
+        # Store information gained from the header
+        self.__header = header
+        if self.__set_attrs_from_header:
```

Comment:
A bit ugly, admittedly. Probably the better approach is to make the convenience attributes properties and check for existence there.